plugin/core/types.py

```diff
@@ -214,6 +215,75 @@ class ClientStates:
     STOPPING = 2
 
 
+class DocumentFilter:
+    """
+    A document filter denotes a document through properties like language, scheme or pattern. An example is a filter
+    that applies to TypeScript files on disk. Another example is a filter that applies to JSON files with name
+    package.json:
+
+        { "language": "typescript", scheme: "file" }
+        { "language": "json", "pattern": "**/package.json" }
+
+    Sublime Text doesn't understand what a language ID is, so we have to maintain a global translation map from language
+    IDs to selectors. Sublime Text also has no support for patterns. We use the wcmatch library for this.
+    """
+
+    __slots__ = ("language", "scheme", "pattern", "selector", "feature_selector")
+
+    def __init__(
+        self,
+        language: Optional[str] = None,
+        scheme: Optional[str] = None,
+        pattern: Optional[str] = None,
+        feature_selector: str = ""
+    ) -> None:
+        self.language = language
+        self.scheme = scheme
+        self.pattern = pattern
+        self.feature_selector = feature_selector
+        if language:
+            # This the connection between VSCode language IDs and ST selectors.
+            lang_id_map = sublime.load_settings("language-ids.sublime-settings")
+            self.selector = lang_id_map.get(language, "source.{}".format(language))  # type: Optional[str]
+        else:
+            self.selector = None
+
+    def matches(self, view: sublime.View) -> bool:
+        """Does this filter match the view? An empty filter matches any view."""
+        if self.selector:
+            if not view.match_selector(0, self.selector):
+                return False
+        if self.scheme:
+            # Can be "file" or "untitled"?
+            pass
+        if self.pattern:
+            # TODO: use facelessuser's wcmatch library here
+            pass
+        return True
+
+    def score_feature(self, view: sublime.View, pt: int) -> int:
+        return view.score_selector(pt, self.feature_selector)
+
+
+class DocumentSelector:
+    """
+    A DocumentSelector is a list of DocumentFilters. A view matches a DocumentSelector if and only if any one of its
+    filters matches against the view.
+    """
+
+    __slots__ = ("filters",)
+
+    def __init__(self, document_selector: List[Dict[str, Any]]) -> None:
+        self.filters = [DocumentFilter(**document_filter) for document_filter in document_selector]
+
+    def __bool__(self) -> bool:
+        return bool(self.filters)
+
+    def matches(self, view: sublime.View) -> bool:
+        """Does this selector match the view? A selector with no filters matches all views."""
+        return any(filter(lambda f: f.matches(view), self.filters)) if self.filters else True  # type: ignore
```

Comment:
Would this be equivalent without having to ignore types?

```suggestion
        return any(f.matches(view) for f in self.filters) if self.filters else True
```